superscore/backends/test.py

```
@@ -0,0 +1,49 @@
+"""
+Backend that manipulates Entries in-memory for testing purposes.
+"""
+from typing import List
+from uuid import UUID
+
+from superscore.backends.core import _Backend
+from superscore.model import Collection, Entry, Snapshot
+
+
+class TestBackend(_Backend):
+    """Backend that manipulates Entries in-memory, for testing purposes."""
+    def __init__(self, data: List[Entry]):
+        self.data = data
+
+    def save_entry(self, entry: Entry) -> None:
+        if self.get_entry(entry.uuid) is not None:
+            raise ValueError(f"Entry {entry.uuid} already exists")
+        else:
+            self.data.append(entry)
+
+    def get_entry(self, uuid: UUID) -> Entry:
+        stack = self.data.copy()
+        while len(stack) > 0:
+            entry = stack.pop()
+            if entry.uuid == uuid:
+                return entry
+            if isinstance(entry, Collection) or isinstance(entry, Snapshot):
+                stack.extend(entry.children)
+        return None
```

Maybe an actual real question: did we decide on how to handle missing entries in the general case? Should we raise an exception here? We certainly don't need to explicitly `return None`

---

in my other PR I've been returning None, so I think that's my vote

---

I think either way is fine- code that calls this simply needs to know what to expect. So we can either raise an exception right away to force calling code to handle it properly (or perish), or we can return None to avoid an automatic perish but expect the caller to check and handle it themselves, else they get some sort of "Nonetype does not have attribute blah" sort of error.

---

I like how raising an exception parallels the other methods and I think I prefer the automatic perish, but I'm fine with either.

Looking at #25, `FilestoreBackend.get_entry()` uses `dict.get()`, which raises an error if the UUID isn't in the cache. Where you referring to returning `None` somewhere else?

---

I think I also prefer the exception raising, if only because "Nonetype cannot do thing" errors are always a bit annoying to handle, versus "I have to catch the raised SuperScoreNoEntry exception"

---

uuh apparently I'm just hallucinating.  Or your future arguments convinced past me 🤷